runtime/onert/core/src/backend/controlflow/kernel/PermuteLayer.cc

```diff
@@ -217,36 +230,54 @@ void PermuteLayer::run()
     auto src = *src_it;
     auto dst = *dst_it;
 
-    if (src != dst)
+    if (src->total_size() == 0)
     {
-      // Conditions to run permutation with multithreading
-      // 1. The tasks for multithreathing was created
-      // 2. The tasks's size > 1
-      // 3. Both tensors are not dynamic
-      if (_tasks_map.find(src) == _tasks_map.end() || _tasks_map.at(src).size() == 1 ||
-          src->is_dynamic() || dst->is_dynamic())
-      {
-        permute(src, dst, src->num_dimensions());
-      }
-      else
+      assert(dst->total_size() == 0);
+    }
+    else
+    {
+      if (src != dst)
       {
-        auto fn = [&](backend::ITensor &) {
-          dst->access([&](backend::ITensor &) {
-            assert(src->getShape().num_elements() * ir::sizeOfDataType(src->data_type()) <=
-                   src->total_size());
-            assert(src->getShape().num_elements() * ir::sizeOfDataType(src->data_type()) <=
-                   dst->total_size());
-            std::vector<PermuteWorkerTask> &tasks = _tasks_map.at(src);
-            for (size_t i = 0; i < tasks.size(); ++i)
-            {
-              tasks.at(i).setBuffers(src->buffer(), dst->buffer());
-            }
-            assert(tasks.size() >= 1);
-            _external_context->ruy_context()->mutable_thread_pool()->Execute(tasks.size(),
-                                                                             tasks.data());
-          });
-        };
-        src->access(fn);
+        // Conditions to run permutation with multithreading
+        // 1. The tasks for multithreathing was created
+        // 2. The tasks's size > 1
+        // 3. Both tensors are not dynamic
+        if (_tasks_map.find(src) == _tasks_map.end() || _tasks_map.at(src).size() == 1 ||
+            src->is_dynamic() || dst->is_dynamic())
+        {
+          permute(src, dst, src->num_dimensions());
+        }
+        // If dst is subtensor, we have to use clEnqueueMapBuffer instead of clEnqueueWirteBuffer
+        else if (dst->needMemoryMap() && !dst->is_subtensor())
+        {
+          if (!src->has_padding() && !dst->has_padding() && src->layout() == dst->layout())
+          {
+            // This is more effective than multi-threading
+            src->access([&](backend::ITensor &) { dst->enqueueWriteBuffer(src->buffer(), false); });
+          }
+          else
+          {
+            _buffers_map[dst].reserve(dst->total_size());
+            auto dst_buffer = _buffers_map[dst].data();
+
+            src->access([&](backend::ITensor &) { runPermuteTasks(src, dst_buffer); });
+            dst->enqueueWriteBuffer(dst_buffer, false);
```

Comment:
This block can be slower than calling ` runPermuteTasks(src, dst->buffer());` if padding size is big.